packages/http-client/src/http-client.ts

```diff
@@ -94,22 +90,109 @@ export function addDestinationToRequestConfig<T extends HttpRequestConfig>(
  * @internal
  */
 export function execute<ReturnT>(executeFn: ExecuteHttpRequestFn<ReturnT>) {
-  return async function <T extends HttpRequestConfig>(
+  return async function <T extends HttpRequestConfigWithOrigin>(
     destination: Destination | DestinationFetchOptions,
     requestConfig: T,
     options?: HttpRequestOptions
   ): Promise<ReturnT> {
+    const resolvedDestination = await resolveDestination(destination);
+    if (!resolvedDestination) {
+      throw Error(
+        `Failed to resolve the destination '${toDestinationNameUrl(
+          destination
+        )}'.`
+      );
+    }
     const destinationRequestConfig = await buildHttpRequest(
-      destination,
-      requestConfig.headers
+      resolvedDestination
     );
-    const request = merge(destinationRequestConfig, requestConfig);
+    logCustomHeadersWarning(requestConfig.headers?.custom);
+    const request = await buildRequestWithMergedHeadersAndQueryParameters(
+      requestConfig,
+      resolvedDestination,
+      destinationRequestConfig
+    );
+
     request.headers = await addCsrfTokenToHeader(destination, request, options);
     logRequestInformation(request);
     return executeFn(request);
   };
 }
 
+/**
+ * @internal
+ * Build a request config from a given request config and a destination.
+ * In addition to merging the information from the request config and the destination, it also picks values with higher priority for headers and query parameters.
+ * @param requestConfig - Any object representing an HTTP request.
+ * @param destination - A resolved [[Destination]] object.
+ * @param destinationRequestConfig - A [[DestinationHttpRequestConfig]] object, that is built from a [[Destination]].
+ * @see [[mergeOptionsWithPriority]]
+ * @returns A resulting request config.
+ */
+export async function buildRequestWithMergedHeadersAndQueryParameters(
+  requestConfig: HttpRequestConfigWithOrigin,
+  destination: Destination,
+  destinationRequestConfig: DestinationHttpRequestConfig
+): Promise<HttpRequestConfig & DestinationHttpRequestConfig> {
+  const { paramsOriginOptions, headersOriginOptions, requestConfigBase } =
+    splitRequestConfig(requestConfig);
+  const headersDestination = await buildHeaders(destination);
+  const headersAndQueryParametersDestinationProperty: Pick<
+    Destination,
+    'headers' | 'queryParameters'
+  > = getAdditionalHeadersAndQueryParameters(
+    (destination.originalProperties as DestinationConfiguration) || {}
+  );
+
+  const mergedQueryParameter = mergeOptionsWithPriority({
```

Comment:
Here a lot is going on, perhaps you can wrap it in two parts for query parameter and header? This would also suggest a split of `headersAndQueryParametersDestinationProperty` in the two cases.